lib/src/shared/main/java/com/couchbase/lite/internal/core/C4NativePeer.java

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright (c) 2020 Couchbase, Inc All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+package com.couchbase.lite.internal.core;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.couchbase.lite.LiteCoreException;
+import com.couchbase.lite.LogDomain;
+import com.couchbase.lite.internal.CouchbaseLiteInternal;
+import com.couchbase.lite.internal.support.Log;
+import com.couchbase.lite.internal.utils.Preconditions;
+import com.couchbase.lite.utils.Fn;
+
+
+/**
+ * This approach exposes the native handle, because anybody can call `get`.
+ */
+public abstract class C4NativePeer extends AtomicLong {
```

Comment:
I couldn't follow why it extends AtomicLong?

Reply:
Well... the goal is to reduce the likelihood of an operation on a freed object.  That means you need to be able to get the handle and set it to 0 atomically. There are two ways to do that: `synchronized` or an AtomicLong.  The bad news is that in order to make the code consistent, it should be possible to use the atomic get/set from the finalizer.  In the finalizer, you cannot use data members because there is no guarantee of finalizer ordering: the data member might be freed first.  That limits the possibilities to either 'synchronized(this)` or subclassing `AtomicLong`.  Doing the former seemed more brittle, to me, because, if any other code happened to be locking on the object, the `synchronized(this)` would block finalization (which seem to be where a lot of our memory goes: objects waiting for finalization).  Subclassing `AtomicLong` guarantees that none of that can happen.  The downside, of course, is that, right now, anybody call call C4XXX.set() or worse, C4XXX.set().  Again, I expect to fix that in a subsequent change.

Reply:
Note, btw, that the lock strategy *would* completely eliminate the race condition, unless the code leaks the peer handle.  ... which it does, so there is nothing to be gained, by using a lock, as the code stands, now.

Reply:
Got it now. I guessed alternatively is to put the handle (AtomicLong) as a variable in the C4NativePeer so the set() method will not be exposed.

Reply:
Right... but that won't work.  When the object is finalized, if the reference to the AtomicLong is in a field of C4NativePeer, then it may be GCed first!

Reply:
Umm... GC + Finalizer are so complicated!

Reply:
Indeed...